Source/EasyNetQ.DI.Tests/ContainerAdapterTests.cs

```diff
@@ -35,92 +24,229 @@ public void Should_last_registration_win(ResolverFactory resolverFactory)
+    {
+        var first = new Service();
+        var last = new Service();
+
+        var resolver = resolverFactory(c =>
+        {
+            c.Register<IService>(first);
+            c.TryRegister<IService>(last);
+        });
+
+        resolver.Resolve<IService>().Should().Be(first);
     }
 
     [Theory]
-    [MemberData(nameof(GetContainerAdapters))]
-    public void Should_singleton_created_once(ResolverFactory resolverFactory)
+    [ClassData(typeof(ContainerAdaptersData))]
+    public void Should_first_registration_win_type(string name, ResolverFactory resolverFactory)
+    {
+        var resolver = resolverFactory(c =>
+        {
+            c.Register<IService, Service>();
+            c.TryRegister<IService, DummyService>();
+        });
+
+        resolver.Resolve<IService>().Should().BeOfType<Service>();
+    }
+
+    [Theory]
+    [ClassData(typeof(ContainerAdaptersData))]
+    public void Should_first_registration_win_factory(string name, ResolverFactory resolverFactory)
+    {
+        var resolver = resolverFactory(c =>
+        {
+            c.Register<IService, Service>(resolver => new Service());
+            c.TryRegister<IService, DummyService>(r => new DummyService());
+        });
+
+        resolver.Resolve<IService>().Should().BeOfType<Service>();
+    }
+
+    [Theory]
+    [ClassData(typeof(ContainerAdaptersData))]
+    public void Should_allow_multiple_try_register_instance(string name, ResolverFactory resolverFactory)
+    {
+        var first = new Service();
+        var last = new Service();
+
+        var resolver = resolverFactory(c =>
+        {
+            c.TryRegister<IService>(first);
+            c.TryRegister<IService>(last);
+            c.TryRegister<IService>(last);
+        });
+
+        resolver.Resolve<IService>().Should().Be(first);
+    }
+
+    [Theory]
+    [ClassData(typeof(ContainerAdaptersData))]
+    public void Should_allow_multiple_try_register_type(string name, ResolverFactory resolverFactory)
+    {
+        var first = new Service();
+        var last = new Service();
+
+        var resolver = resolverFactory(c =>
+        {
+            c.TryRegister<IService>(first);
+            c.TryRegister<IService>(last);
+        });
+
+        resolver.Resolve<IService>().Should().Be(first);
+    }
+
+    [Theory]
+    [ClassData(typeof(ContainerAdaptersData))]
+    public void Should_allow_multiple_try_register_factory(string name, ResolverFactory resolverFactory)
+    {
+        var resolver = resolverFactory(c =>
+        {
+            c.TryRegister<IService, Service>();
+            c.TryRegister<IService, DummyService>();
+        });
+
+        resolver.Resolve<IService>().Should().BeOfType<Service>();
+    }
+
+    [Theory]
+    [ClassData(typeof(ContainerAdaptersData))]
+    public void Should_resolve_single_registration(string name, ResolverFactory resolverFactory)
+    {
+        if (name == "Autofac")
```

Comment:
Still here? You said thst you fixed Autofac registrations.

Reply:
Yep. 

It is the only thing left(I've fixed TryRegister methods for Autofac, because it supports conditional registration). There is only one way at the moment to resolve this: to copy container builder each time when Register is called.

Reply:
OK. May be post "known issue" about such Autofac design?